datalad/utils.py

```diff
@@ -117,6 +117,165 @@ def rmtree(path, chmod_files='auto', *args, **kwargs):
     rotree(path, ro=False, chmod_files=chmod_files)
     shutil.rmtree(path, *args, **kwargs)
 
+
+def is_symlink(file_, good=None):
+    """Check if file_ is a symlink
+
+    Parameters
+    ----------
+    good: None or bool, optional
+        Test not only that it is a symlink, but either it is a good symlink
+        (good=True) or broken (good=False) symlink.  So if good=False,
+        it would return True only if it is a symlink and it is broken.
+    """
+    try:
+        link = os.readlink(file_)
+    except OSError:
+        link = None
+
+    # TODO: copied from ok_symlink so may be we also want to check the path etc?
+    # path_ = realpath(file_)
+    # ok_(path_ != link)
+    # TODO anything else?
+
+    if good is None:
+        # nothing to do more:
+        return bool(link)
+    else:
+        return is_good_symlink(file_)
+
+
+def is_good_symlink(file_):
+    """Return if a file_ is a good symlink (assuming it is a symlink)
+    """
+    return exists(realpath(file_))
+
+def ls_tree(path, files_only=False, relative=True):
+    """Recurse into the directory and return all found files (and dirs if dirs)
+    """
+    out = []
+    for root, dirs, files in os.walk(path):
+        for f in files:
+            out.append(opj(root, f))
+        if not files_only:
+            for d in dirs:
+                out.append(opj(root, d))
+    out = sorted(out)
+    if relative:
+        beg = len(path) + len(os.pathsep)
+        return [o[beg:] for o in out]
+    else:
+        return out
+
+def has_content(file_):
+    """Returns True if a file is not a broken symlink, and if it has content
+    """
+    if not on_windows:
+        if not is_symlink(file_, good=True):
+            return False
+    # verify it has content
+    return os.stat(realpath(file_)).st_size > 0
```

Comment:
Yeah - your test working on that with_tree, where you create empty files.
Thes are empty on windows too of course.
And remember: It's not really windows, but direct mode.
May be the best way is to enhance is_symlink to detect whether it is a
"git-annex direct mode symlink" as well, by looking into the file and check
whether it is a single line containing a path to .git/annex/ .....

2015-04-07 16:11 GMT+02:00 Yaroslav Halchenko <notifications@github.com>:

> In datalad/utils.py
> <https://github.com/datalad/datalad/pull/101#discussion_r27880866>:
>
> > +                out.append(opj(root, d))
> > +    out = sorted(out)
> > +    if relative:
> > +        beg = len(path) + len(os.pathsep)
> > +        return [o[beg:] for o in out]
> > +    else:
> > +        return out
> > +
> > +def has_content(file_):
> > +    """Returns True if a file is not a broken symlink, and if it has content
> > +    """
> > +    if not on_windows:
> > +        if not is_symlink(file_, good=True):
> > +            return False
> > +    # verify it has content
> > +    return os.stat(realpath(file_)).st_size > 0
>
> damn, right... so for actual "annex" repositories it would be the case and
> things will not work as expected. Thus I would need to fulfill that TODO
> item to check with annex about file availability (instead of just a
> symlink/empty), unless there is some weird Windows specific Python function
> in os. which could tell me if that is a "symlink" file (probably not, and
> it is just a git way to overcome the problem of absent symlinks?)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/datalad/datalad/pull/101/files#r27880866>.
>